controllers/servicefence_controller.go

```diff
@@ -282,6 +344,106 @@ func newPrometheusSourceConfig(env bootstrap.Environment) (metric.PrometheusSour
 	}, nil
 }
 
+func accessLogHandler(logEntry []*data_accesslog.HTTPAccessLogEntry, clientSet *kubernetes.Clientset) (map[string]map[string]string, error) {
+	log := log.WithField("reporter", "accesslog convertor handler")
+	tmpResult := make(map[string]map[string]string)
+	for _, entry := range logEntry {
+		tmpValue := make(map[string]string)
+
+		// fetch sourceEp
+		sourceEp, err := fetchSourceEp(entry, log)
+		if err != nil {
+			return nil, err
+		}
+		if sourceEp == "" {
+			continue
+		}
+
+		// fetch sourceSvcMeta
+		sourceSvcMeta, err := fetchSourceSvcMeta(sourceEp, clientSet, log)
+		if err != nil {
+			return nil, err
+		}
+		if sourceSvcMeta == "" {
+			continue
+		}
+
+		// fetch destinationSvcMeta
+		destinationSvc := fetchDestinationSvc(entry, log)
+		if destinationSvc == "" {
+			continue
+		}
+
+		// push result
+		if _, ok := tmpResult[sourceSvcMeta]; !ok {
+			tmpValue[destinationSvc] = "1"
+			tmpResult[sourceSvcMeta] = tmpValue
+		} else if _, ok = tmpResult[sourceSvcMeta][destinationSvc]; !ok {
+			tmpResult[sourceSvcMeta][destinationSvc] = "1"
+		} else {
+			v, err := strconv.Atoi(tmpResult[sourceSvcMeta][destinationSvc])
+			if err != nil {
+				return nil, err
+			}
+			tmpResult[sourceSvcMeta][destinationSvc] = strconv.Itoa(v + 1)
+		}
+		log.Debugf("tmpResult[%s][%s]: %s", sourceSvcMeta, destinationSvc, tmpResult[sourceSvcMeta][destinationSvc])
+	}
+
+	return tmpResult, nil
+}
+
+func fetchSourceEp(entry *data_accesslog.HTTPAccessLogEntry, log *logrus.Entry) (string, error) {
+	if entry.CommonProperties.DownstreamRemoteAddress == nil {
+		log.Debugf("DownstreamRemoteAddress is nil, skip")
+		return "", nil
+	}
+	downstreamSock, ok := entry.CommonProperties.DownstreamRemoteAddress.Address.(*envoy_config_core.Address_SocketAddress)
+	if !ok {
+		return "", stderrors.New("wrong type of DownstreamRemoteAddress")
+	}
+	if downstreamSock == nil || downstreamSock.SocketAddress == nil {
+		return "", stderrors.New("downstream socket address is nil")
+	}
+	log.Debugf("SourceEp is: %s", downstreamSock.SocketAddress.Address)
+	return downstreamSock.SocketAddress.Address, nil
+}
+
+func fetchSourceSvcMeta(sourceEp string, clientSet *kubernetes.Clientset, log *logrus.Entry) (string, error) {
+	eps, err := clientSet.CoreV1().Endpoints("").List(metav1.ListOptions{})
```

Comment:
expensive!

Reply:
add function generateSvcToIpsCache, and use this cache and rwmutex to splice SourceSvc